docs/dev_plan.md

```diff
@@ -0,0 +1,339 @@
+changing a solver type without modifying the modules around it.
+The flow of data should be considered in discrete steps in a pipeline
+rather than a monolithic system working on the data.
+
+```{mermaid}
+flowchart LR
+
+    Start(( ))
+
+    subgraph I/O
+        direction LR
+        io1{{YAML}}
+        io2{{JSON}}
+    end
+
+    DataModel[[Data Model]]
+
+    subgraph Solver
+        direction LR
+        solver1{{Type 1}}
+        solver2{{Type 2}}
+        solver3{{Type 3}}
+    end
+
+    Output[[Output Model]]
+
+    subgraph Export
+        direction LR
+        export1{{Visualization}}
+        export2{{ASCII}}
+        export3{{Commercial}}
+    end
+
+    Finish(( ))
+
+    Start --- I/O
+    I/O --- DataModel
+    DataModel --- Solver
+    Solver --- Output
+    Output --- Export
+    Export --- Finish
+```
+
+It is critical to the sustainability and stability of OpenTurbine
+to maintain independence from external software even though there
+will be reliance on existing libraries for common tasks. The modular
+design should include data structures and API's that are general enough
+to support integration of third party libraries as well as the ability
+to change any library for another that accomplishes a similar task
+even if by alternative methods.
+
+### Performance first
+
+A key design consideration of OpenTurbine is computational efficiency
+or performance. Both the quantity of work and the efficiency of data should
+be considered and measured (i.e., profiled) during any development effort.
+A modular architecture should support offloading computationally expensive
+tasks to hardware accelerators or specialized libraries, and support
+multiple options for doing so depending on user configurations and
+the computational environment.
+
+Similar to modularity in the architecture, expensive tasks should be
+structured in a kernel form. This low-level design pattern combines
+expensive mathematical operations into an aggregate form, and structures
+them so that performance libraries or compilers can parallelize the
+computation. This pattern helps to encapsulate expensive operations and
+algorithms. Additionally, it follows the modular architecture design
+described above in that it supports swapping accelerators or parallelization
+methods../
+
+
+#### Data-oriented design
+
+OpenTurbine developers designing new algorithms and data structures
+should become familiar with the concepts of [data-oriented design](https://en.wikipedia.org/wiki/Data-oriented_design),
+particularly [structures of arrays vs. arrays of structures](https://stackoverflow.com/questions/17924705/structure-of-arrays-vs-array-of-structures).
+The key concept of this paradigm is to structure data so that it maps
+closely to the form it will be represented and used within the relevant
+processing units. For example, it is a common mistake in object-oriented
+design to represent data as arrays of structures (AoS),
```

Comment:
It's not always a mistake. The preference for SoA vs AoS depends on the most computationally expensive algorithm, and which order the data needs to be accessed.

Reply:
Good point @alanw0. In addition to being inaccurate, a description of the idea of data-oriented design is out of scope for this document. I've made some improvements to this section and removed much of the explanation based on your feedback.